lib/isaac.hpp

```diff
@@ -68,1792 +69,1948 @@ template <
+                                                                      )
+                        )
+                    )
+                );
+                pointer_array.pointer[I] = alpaka::mem::view::getPtrNative ( alpaka_vector.back() );
+#else
+                ISAAC_CUDA_CHECK ( cudaMalloc ( ( void** ) & ( pointer_array.pointer[I] ), sizeof ( isaac_float_dim< TSource::feature_dim > ) * ( local_size[0] + 2 * ISAAC_GUARD_SIZE ) * ( local_size[1] + 2 * ISAAC_GUARD_SIZE ) * ( local_size[2] + 2 * ISAAC_GUARD_SIZE ) ) );
 #endif
-                        {
-                            grid_size.x = ISAAC_IDX_TYPE(local_size[0] + ISAAC_GUARD_SIZE * 2);
-                            grid_size.y = ISAAC_IDX_TYPE(local_size[0] + ISAAC_GUARD_SIZE * 2);
-                            block_size.x = ISAAC_IDX_TYPE(1);
-                            block_size.y = ISAAC_IDX_TYPE(1);
-                        }
-                        const alpaka::vec::Vec<TAccDim, ISAAC_IDX_TYPE> threads (ISAAC_IDX_TYPE(1), ISAAC_IDX_TYPE(1), ISAAC_IDX_TYPE(1));
-                        const alpaka::vec::Vec<TAccDim, ISAAC_IDX_TYPE> blocks  (ISAAC_IDX_TYPE(1), block_size.x, block_size.y);
-                        const alpaka::vec::Vec<TAccDim, ISAAC_IDX_TYPE> grid    (ISAAC_IDX_TYPE(1), grid_size.x, grid_size.y);
-                        auto const workdiv(alpaka::workdiv::WorkDivMembers<TAccDim, ISAAC_IDX_TYPE>(grid,blocks,threads));
-                        updateBufferKernel<TSource> kernel;
-                        auto const instance
-                        (
-                            alpaka::exec::create<TAcc>
-                            (
-                                workdiv,
-                                kernel,
-                                source,
-                                pointer_array.pointer[ I ],
-                                local_size_array
-                            )
-                        );
-                        alpaka::queue::enqueue(stream, instance);
-                        alpaka::wait::wait(stream);
-                    #else
-                        dim3 block (block_size.x, block_size.y);
-                        dim3 grid  (grid_size.x, grid_size.y);
-                        updateBufferKernel<<<grid,block>>>( source, pointer_array.pointer[ I ], local_size_array );
-                        ISAAC_CUDA_CHECK(cudaDeviceSynchronize());
-                    #endif
-                }
             }
-        };
-
-        struct calc_minmax_iterator
+        }
+    };
+
+    struct update_particle_source_iterator {
+        template
+        <
+            typename TParticleSource,
+            typename TWeight,
+            typename TOffset,
+            typename TPointer
+            >
+        ISAAC_HOST_INLINE  void operator() (
+            const int I,
+            TParticleSource& particle_source,
+            const TWeight& weight,
+            const TOffset& weightArrayOffset,
```

Comment:
could you please add a doxygen string to the function and give a short description what the arguments mean.